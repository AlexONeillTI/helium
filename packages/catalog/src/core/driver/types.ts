import { GlobalTypes, CatalogContentLazyQueryHookResult } from '@thoughtindustries/content';

export type AggregationFilter = {
  label: string;
  value: string;
};

export enum SortField {
  UpdatedAt = 'updatedAt',
  CreatedAt = 'createdAt',
  Title = 'title',
  PublishedDate = 'publishDate',
  CourseStartDate = 'courseStartDate',
  Relevance = 'relevance'
}

export enum SortDirection {
  Desc = 'desc',
  Asc = 'asc'
}

export type Sort = {
  field: SortField;
  direction?: SortDirection;
};

export type CatalogDriverRequestURLState = {
  page?: number;
  searchTerm?: string;
  aggregationFilters: AggregationFilter[];
  token?: string;
};

export type CatalogDriverRequestState = CatalogDriverRequestURLState & {
  sort?: Sort;
  displayType?: GlobalTypes.ContentItemDisplayType;
  contentTypes: GlobalTypes.ContentKind[];
};

export type CatalogDriverResponseState = {
  results: GlobalTypes.Content[];
  queryCustomFields?: JSON;
  aggregations: GlobalTypes.Aggregation[];
  total?: number;
  hasMore: boolean;
  isCurated: boolean;
  tokenLabel?: string;
  resultsDisplayType?: GlobalTypes.ContentItemDisplayType;
  enabledSorts: Sort[];
  enabledDisplayTypes: GlobalTypes.ContentItemDisplayType[];
  resultContentTypes: GlobalTypes.ContentKind[];
  contentTypeFilterEnabled: boolean;
  displayBundle?: GlobalTypes.Bundle;
  // TODO: check if the followings can be handled on server side
  displayStartDateEnabled: boolean;
  displayAuthorsEnabled: boolean;
  // TODO: check if this is missing from schema
  displayDescriptionOnCalendar: boolean;
};

export type CatalogDriverGeneralState = {
  error?: string;
  isLoading: boolean;
  // TODO: check if this could become request state once API supports it
  pageSize: number;
};

export type CatalogDriverState = CatalogDriverRequestState &
  CatalogDriverResponseState &
  CatalogDriverGeneralState;

export type CatalogDriverConfig = {
  layoutId?: string;
  widgetId?: string;
  onSearch: CatalogContentLazyQueryHookResult[0];
  initialState?: Partial<CatalogDriverRequestState>;
  trackUrlState?: boolean;
  alwaysSearchOnInitialLoad?: boolean;
};

type CatalogDriverUpdateResultsOptions = {
  skipPushToUrl?: boolean;
  replaceUrl?: boolean;
};
export type CatalogDriverUpdateResultsFn = (
  requestParams: Partial<CatalogDriverState>,
  options?: CatalogDriverUpdateResultsOptions
) => Promise<void>;
export type CatalogDriverSubscriptionFn = (state: CatalogDriverState) => void;

export type CatalogDriverMakeRequestFn = (
  options: CatalogDriverUpdateResultsOptions
) => Promise<void>;
