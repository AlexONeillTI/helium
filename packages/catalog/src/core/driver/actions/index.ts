export { default as addAggregationFilter } from './add-aggregation-filter';
export { default as addContentType } from './add-content-type';
export { default as removeAggregationFilter } from './remove-aggregation-filter';
export { default as removeContentType } from './remove-content-type';
export { default as removeSearchTerm } from './remove-search-term';
export { default as removeToken } from './remove-token';
export { default as setDisplayType } from './set-display-type';
export { default as setPage } from './set-page';
export { default as setSearchTerm } from './set-search-term';
export { default as setSort } from './set-sort';

import { Fn as addAggregationFilterFn } from './add-aggregation-filter';
import { Fn as addContentTypeFn } from './add-content-type';
import { Fn as removeAggregationFilterFn } from './remove-aggregation-filter';
import { Fn as removeContentTypeFn } from './remove-content-type';
import { Fn as removeSearchTermFn } from './remove-search-term';
import { Fn as removeTokenFn } from './remove-token';
import { Fn as setDisplayTypeFn } from './set-display-type';
import { Fn as setPageFn } from './set-page';
import { Fn as setSearchTermFn } from './set-search-term';
import { Fn as setSortFn } from './set-sort';
export type CatalogDriverActions = {
  addAggregationFilter: addAggregationFilterFn;
  addContentType: addContentTypeFn;
  removeAggregationFilter: removeAggregationFilterFn;
  removeContentType: removeContentTypeFn;
  removeSearchTerm: removeSearchTermFn;
  removeToken: removeTokenFn;
  setDisplayType: setDisplayTypeFn;
  setPage: setPageFn;
  setSearchTerm: setSearchTermFn;
  setSort: setSortFn;
};
