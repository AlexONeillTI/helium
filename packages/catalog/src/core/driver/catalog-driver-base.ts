import { CatalogContentQueryVariables } from '@thoughtindustries/content';
import { ContentKind } from '@thoughtindustries/content/src/graphql/global-types';
import { DEFAULT_PAGE, DEFAULT_PAGE_SIZE } from './constants';
import {
  CatalogDriverConfig,
  CatalogDriverMakeRequestFn,
  CatalogDriverState,
  CatalogDriverSubscriptionFn,
  CatalogDriverUpdateResultsFn
} from './types';
import URLManager from './url-manager';
import {
  stateToSearchParams,
  toEnabledDisplayTypes,
  toEnabledSorts,
  shouldUpdateContentTypes
} from './utilities';

const filterSearchParameters = ({
  page,
  aggregationFilters,
  searchTerm,
  sort,
  displayType,
  token,
  contentTypes
}: CatalogDriverState) => ({
  page,
  aggregationFilters,
  searchTerm,
  sort,
  displayType,
  token,
  contentTypes
});

export const DEFAULT_STATE: CatalogDriverState = {
  // default request state
  page: DEFAULT_PAGE,
  aggregationFilters: [],
  contentTypes: [],
  // default response state
  results: [],
  aggregations: [],
  hasMore: false,
  isCurated: false,
  enabledSorts: [],
  enabledDisplayTypes: [],
  resultContentTypes: [],
  contentTypeFilterEnabled: false,
  displayStartDateEnabled: false,
  displayAuthorsEnabled: false,
  displayDescriptionOnCalendar: false,
  // default general state
  isLoading: false,
  pageSize: DEFAULT_PAGE_SIZE
};

/**
 * Base class for CatalogDriver
 */
export default class CatalogDriverBase {
  private _initialized = false;
  private _state: CatalogDriverState = DEFAULT_STATE;
  private readonly _onSearch;
  private _subscriptions: CatalogDriverSubscriptionFn[];
  private readonly _layoutId;
  private readonly _widgetId;
  private readonly _trackUrlState;
  private readonly _urlManager: URLManager | undefined;
  private readonly _alwaysSearchOnInitialLoad;
  private _prevSearchParams: CatalogContentQueryVariables | undefined = undefined;

  constructor({
    initialState,
    onSearch,
    layoutId,
    widgetId,
    trackUrlState = true,
    alwaysSearchOnInitialLoad = false
  }: CatalogDriverConfig) {
    this._onSearch = onSearch;
    this._subscriptions = [];
    this._layoutId = layoutId;
    this._widgetId = widgetId;
    this._trackUrlState = trackUrlState;
    this._alwaysSearchOnInitialLoad = alwaysSearchOnInitialLoad;

    let urlState = {};
    if (trackUrlState) {
      this._urlManager = new URLManager();
      urlState = this._urlManager.getStateFromURL();
      this._urlManager.onURLStateChange(urlState => {
        this.updateResults({ ...urlState }, { skipPushToUrl: true });
      });
    }

    // We filter these here to disallow anything other than valid search
    // parameters to be passed in initial state, or url state. `results`, etc,
    // should not be allowed to be passed in, that should be generated based on
    // the results of the query
    const searchParameters = filterSearchParameters({
      ...this._state,
      ...initialState,
      ...urlState
    });

    // Initialize the state without calling setState, because we don't
    // want to trigger an update callback, we're just initializing the state
    // to the correct default values for the initial UI render
    this._state = {
      ...this._state,
      ...searchParameters
    };
  }

  /**
   * After creating new instance of the CatalogDriver and before triggering
   * any action, will call this to initialize some resources (Like: trigger
   * initial search).
   */
  init = async () => {
    if (this._initialized) {
      return;
    }

    this._initialized = true;

    const searchParameters = filterSearchParameters(this._state);

    // We'll trigger an initial search if initial parameters contain a search term,
    // aggregation filters or token, or if alwaysSearchOnInitialLoad is set.
    // Otherwise, we'll just save their selections in state as initial values.
    if (
      searchParameters.searchTerm ||
      searchParameters.aggregationFilters.length > 0 ||
      searchParameters.token ||
      this._alwaysSearchOnInitialLoad
    ) {
      await this.updateResults(searchParameters, { replaceUrl: true });
    }
  };

  /**
   * Exposed to actions to update results. We want to leverage
   * the `isLoading` state to limit one server request in flight at a
   * time. It serves as the stop gate to validate the action request.
   * Once validated, it will flow to update state as well as making the
   * server request. We also want to make sure the instance should be
   * initialized first, otherwise actions will be discarded.
   *
   * @param requestParams - request state to be merged with current state for making the server reqest
   * @param options - determine how url tracking is handled after server request
   */
  protected updateResults: CatalogDriverUpdateResultsFn = async (
    requestParams,
    { skipPushToUrl, replaceUrl } = {}
  ) => {
    // Skip if not initialized
    if (!this._initialized) {
      return;
    }

    // Skip when a server request is in-flight
    const { isLoading } = this._state;
    if (isLoading) {
      return;
    }

    // Extract request state
    const {
      page,
      pageSize,
      aggregationFilters,
      searchTerm,
      sort,
      displayType,
      token,
      contentTypes
    } = {
      ...this._state,
      ...requestParams
    };

    // update state
    this.setState({
      page,
      pageSize,
      aggregationFilters,
      searchTerm,
      sort,
      displayType,
      token,
      contentTypes
    });

    await this._makeRequest({
      skipPushToUrl,
      replaceUrl
    });
  };

  /**
   * This method is continued from updateResults.
   *
   * Application state updates from actions are performed in updateResults.
   * The state will be updated after the API calls.
   *
   * @param {Object} options - determine how url tracking is handled after server request
   */
  private _makeRequest: CatalogDriverMakeRequestFn = async ({ skipPushToUrl, replaceUrl }) => {
    this.setState({
      isLoading: true
    });

    const searchParams = stateToSearchParams(this._state);
    // check if there is custom settings
    if (this._layoutId && this._widgetId) {
      searchParams.layoutId = this._layoutId;
      searchParams.widgetId = this._widgetId;
    }
    const { data, error } = await this._onSearch({
      variables: searchParams
    });

    this.setState({
      isLoading: false
    });

    if (error || !data) {
      this.setState({
        error: `An unexpected error occurred: ${error ? error.message : 'empty data'}`
      });
      return;
    }

    const {
      meta: {
        displayBundle,
        tokenLabel,
        total,
        hasMore,
        isCurated,
        aggregations,
        contentTypes = [],
        resultsDisplayType,
        sortUpdatedAtEnabled,
        sortCreatedAtEnabled,
        sortTitleEnabled,
        sortPublishDateEnabled,
        sortCourseStartDateEnabled,
        sortRelevanceEnabled,
        displayTypeListEnabled,
        displayTypeGridEnabled,
        displayTypeCalendarEnabled,
        displayStartDateEnabled,
        displayAuthorsEnabled,
        displayDescriptionOnCalendar,
        contentTypeFilterEnabled,
        queryCustomFields
      },
      contentItems
    } = data.CatalogContent;

    // transform result meta to enabled sorts
    const enabledSorts = toEnabledSorts({
      sortUpdatedAtEnabled,
      sortCreatedAtEnabled,
      sortTitleEnabled,
      sortPublishDateEnabled,
      sortCourseStartDateEnabled,
      sortRelevanceEnabled
    });

    // transform result meta to enabled display types
    const enabledDisplayTypes = toEnabledDisplayTypes({
      displayTypeListEnabled,
      displayTypeGridEnabled,
      displayTypeCalendarEnabled
    });

    // transform result meta to result content types
    const shouldUpdate = shouldUpdateContentTypes(searchParams, this._prevSearchParams);
    const { resultContentTypes: currentContentTypes } = this._state;
    let resultContentTypes = [...currentContentTypes];
    if (shouldUpdate || currentContentTypes.length < contentTypes.length) {
      resultContentTypes = [...contentTypes] as ContentKind[];
    }

    this.setState({
      error: undefined,
      results: contentItems,
      queryCustomFields,
      aggregations,
      total,
      hasMore,
      isCurated,
      tokenLabel,
      resultsDisplayType,
      enabledSorts,
      enabledDisplayTypes,
      resultContentTypes,
      displayStartDateEnabled,
      displayAuthorsEnabled,
      displayDescriptionOnCalendar,
      contentTypeFilterEnabled,
      displayBundle
    });

    this._prevSearchParams = { ...searchParams };

    // (optional) push state to url
    if (!skipPushToUrl && this._trackUrlState && this._urlManager) {
      const { page, aggregationFilters, searchTerm, token } = this._state;
      this._urlManager.pushStateToURL(
        {
          page,
          aggregationFilters,
          searchTerm,
          token
        },
        { replaceUrl }
      );
    }
  };

  /**
   * Any time state is updated in this CatalogDriver, the provided callback
   * will be executed with the updated state.
   *
   * @param onStateChange - callback when state changes
   */
  subscribeToStateChanges(onStateChange: CatalogDriverSubscriptionFn) {
    this._subscriptions.push(onStateChange);
  }

  /**
   * Remove a state update subscriber from this CatalogDriver
   *
   * @param onStateChange - callback when state changes
   */
  unsubscribeToStateChanges(onStateChange: CatalogDriverSubscriptionFn) {
    this._subscriptions = this._subscriptions.filter(sub => sub !== onStateChange);
  }

  /**
   * Tear down resources for this CatalogDriver
   */
  tearDown() {
    this._subscriptions = [];
    this._urlManager && this._urlManager.tearDown();
  }

  getState() {
    return this._state;
  }

  protected setState(newState: Partial<CatalogDriverState>) {
    const state = { ...this._state, ...newState };
    this._state = state;
    this._subscriptions.forEach(subscription => subscription(state));
  }
}
