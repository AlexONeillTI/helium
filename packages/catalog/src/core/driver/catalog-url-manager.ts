import {
  AggregationFilter,
  CatalogDriverRequestState,
  CatalogDriverRequestURLState
} from './types';
import { aggregationFiltersToParams } from './utilities';

type CatalogParsedURL = {
  pathname: string;
  searchString?: string;
};

type AggregationFilterWithComposedURL = {
  filter: AggregationFilter;
  url: string;
};

const toInteger = (str: string): number => parseInt(str, 10);

const parsePageFromQueryParams = (params: URLSearchParams): number | undefined => {
  const page = params.get('page');
  if (!page) {
    return;
  }
  const parsedPage = toInteger(page);
  if (isNaN(parsedPage) || parsedPage < 1) {
    return;
  }
  return parsedPage;
};

// TODO: validate inputs
const parseAggregationFiltersFromQueryParams = (
  params: URLSearchParams
): AggregationFilter[] | undefined => {
  const labels = params.get('labels');
  const values = params.get('values');
  if (!labels || !values) {
    return;
  }
  try {
    const parsedLabels = JSON.parse(labels) as string[];
    const parsedValues = JSON.parse(values) as string[];
    return parsedLabels.reduce((acc, label, index) => {
      if (index in parsedValues && parsedValues[index]) {
        acc.push({ label, value: parsedValues[index] });
      }
      return acc;
    }, [] as AggregationFilter[]);
  } catch {
    return;
  }
};

const parseTokenFromQueryParams = (params: URLSearchParams): string | undefined => {
  const token = params.get('token');
  if (!token) {
    return;
  }
  return token;
};

const paramsToState = (params: URLSearchParams): Partial<CatalogDriverRequestURLState> => {
  const state: Partial<CatalogDriverRequestState> = {};

  const page = parsePageFromQueryParams(params);
  if (page) {
    state.page = page;
  }

  const aggregationFilters = parseAggregationFiltersFromQueryParams(params);
  if (aggregationFilters) {
    state.aggregationFilters = aggregationFilters;
  }

  const token = parseTokenFromQueryParams(params);
  if (token) {
    state.token = token;
  }

  return state;
};

/**
 * The URL Manager is responsible for parsing URL and composing URLs
 * for server side routing. When user interacts with the UI, user will
 * be redirected to the new route reflecting their interactions.
 *
 * Note that only CatalogDriver relevant query params are handled here.
 * For some URL navigations, the `page` param will be reset.
 * For non-relevant query params, composed URLs will carry them over.
 */

export default class CatalogURLManager {
  private readonly _pathname;
  private readonly _searchParams;
  private readonly _parsedState;

  constructor(parsedUrl: CatalogParsedURL) {
    const { pathname, searchString } = parsedUrl;
    this._pathname = pathname;
    this._searchParams = new URLSearchParams(searchString || undefined);
    this._parsedState = paramsToState(this._searchParams);
  }

  private _composeURL(searchString: string): string {
    if (!searchString) {
      return this._pathname;
    }
    return `${this._pathname}?${searchString}`;
  }

  /**
   * Parse the current URL into application state
   *
   * @returns URL state
   */
  getStateFromURL(): Partial<CatalogDriverRequestURLState> {
    return this._parsedState;
  }

  composeURLsForRemoveAggregationFilterBatch(
    filters: AggregationFilter[]
  ): AggregationFilterWithComposedURL[] {
    const { aggregationFilters = [] } = this._parsedState;
    return filters.reduce((acc, currentFilter) => {
      const { label: currentLabel, value: currentValue } = currentFilter;
      const clonedParams = new URLSearchParams(this._searchParams);
      const newFilters = aggregationFilters.filter(
        ({ label, value }) => label !== currentLabel && value !== currentValue
      );

      if (newFilters.length) {
        const transformedFilters = aggregationFiltersToParams(newFilters);
        clonedParams.set('labels', JSON.stringify(transformedFilters.labels));
        clonedParams.set('values', JSON.stringify(transformedFilters.values));
      } else {
        clonedParams.delete('labels');
        clonedParams.delete('values');
      }

      // reset page
      clonedParams.delete('page');
      acc.push({
        filter: currentFilter,
        url: this._composeURL(clonedParams.toString())
      });
      return acc;
    }, [] as AggregationFilterWithComposedURL[]);
  }

  composeURLForAddAggregationFilter(filter: AggregationFilter, isCurated: boolean): string {
    const { aggregationFilters = [] } = isCurated ? {} : this._parsedState;
    const clonedParams = new URLSearchParams(this._searchParams);
    const newFilters = [...aggregationFilters];
    newFilters.push(filter);
    const transformedFilters = aggregationFiltersToParams(newFilters);
    clonedParams.set('labels', JSON.stringify(transformedFilters.labels));
    clonedParams.set('values', JSON.stringify(transformedFilters.values));
    // reset page
    clonedParams.delete('page');
    return this._composeURL(clonedParams.toString());
  }

  composeURLForSetPage(page: number): string {
    const clonedParams = new URLSearchParams(this._searchParams);
    clonedParams.set('page', page.toString());
    return this._composeURL(clonedParams.toString());
  }

  composeURLForRemoveToken(): string {
    const clonedParams = new URLSearchParams(this._searchParams);
    clonedParams.delete('token');
    // reset page
    clonedParams.delete('page');
    return this._composeURL(clonedParams.toString());
  }
}
