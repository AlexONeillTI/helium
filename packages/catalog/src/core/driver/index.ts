export { default as CatalogDriver } from './catalog-driver';
export { DEFAULT_STATE } from './catalog-driver-base';
export type { AggregationFilter, Sort, CatalogDriverState, CatalogDriverConfig } from './types';
export { SortDirection, SortField } from './types';
