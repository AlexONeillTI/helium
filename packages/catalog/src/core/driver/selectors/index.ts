import { GlobalTypes } from '@thoughtindustries/content';
import { CatalogDriverState } from '../types';

export function activeDisplayTypeSelector({
  displayType,
  resultsDisplayType
}: CatalogDriverState): GlobalTypes.ContentItemDisplayType | undefined {
  if (displayType) {
    return displayType;
  }

  return resultsDisplayType;
}

export function activeDisplayTypeIsGridSelector(state: CatalogDriverState): boolean {
  const activeDisplayType = activeDisplayTypeSelector(state);
  return activeDisplayType === GlobalTypes.ContentItemDisplayType.Grid;
}

export function activeDisplayTypeIsListSelector(state: CatalogDriverState): boolean {
  const activeDisplayType = activeDisplayTypeSelector(state);
  return activeDisplayType === GlobalTypes.ContentItemDisplayType.List;
}

export function activeDisplayTypeIsCalendarSelector(state: CatalogDriverState): boolean {
  const activeDisplayType = activeDisplayTypeSelector(state);
  return activeDisplayType === GlobalTypes.ContentItemDisplayType.Calendar;
}
