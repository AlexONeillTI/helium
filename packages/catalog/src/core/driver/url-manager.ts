import { createBrowserHistory as createHistory, BrowserHistory } from 'history';
import {
  AggregationFilter,
  CatalogDriverRequestState,
  CatalogDriverRequestURLState
} from './types';
import { aggregationFiltersToParams } from './utilities';

const toInteger = (str: string): number => parseInt(str, 10);

const parsePageFromQueryParams = (params: URLSearchParams): number | undefined => {
  const page = params.get('page');
  if (!page) {
    return;
  }
  const parsedPage = toInteger(page);
  if (isNaN(parsedPage) || parsedPage < 1) {
    return;
  }
  return parsedPage;
};

const parseAggregationFiltersFromQueryParams = (
  params: URLSearchParams
): AggregationFilter[] | undefined => {
  const labels = params.get('labels');
  const values = params.get('values');
  if (!labels || !values) {
    return;
  }
  try {
    const parsedLabels = JSON.parse(labels) as string[];
    const parsedValues = JSON.parse(values) as string[];
    return parsedLabels.reduce((acc, label, index) => {
      if (index in parsedValues && parsedValues[index]) {
        acc.push({ label, value: parsedValues[index] });
      }
      return acc;
    }, [] as AggregationFilter[]);
  } catch {
    return;
  }
};

const parseSearchTermFromQueryParams = (params: URLSearchParams): string | undefined => {
  const searchTerm = params.get('query');
  if (!searchTerm) {
    return;
  }
  return searchTerm;
};

const parseTokenFromQueryParams = (params: URLSearchParams): string | undefined => {
  const token = params.get('token');
  if (!token) {
    return;
  }
  return token;
};

const paramsToState = (params: URLSearchParams): Partial<CatalogDriverRequestURLState> => {
  const state: Partial<CatalogDriverRequestState> = {};

  const page = parsePageFromQueryParams(params);
  if (page) {
    state.page = page;
  }

  const aggregationFilters = parseAggregationFiltersFromQueryParams(params);
  if (aggregationFilters) {
    state.aggregationFilters = aggregationFilters;
  }

  const searchTerm = parseSearchTermFromQueryParams(params);
  if (searchTerm) {
    state.searchTerm = searchTerm;
  }

  const token = parseTokenFromQueryParams(params);
  if (token) {
    state.token = token;
  }

  return state;
};

const stateToParams = ({
  searchTerm,
  page,
  aggregationFilters,
  token
}: CatalogDriverRequestURLState): URLSearchParams => {
  const params = new URLSearchParams();
  if (page && page > 1) params.set('page', page.toString());
  if (searchTerm) params.set('query', searchTerm);
  if (aggregationFilters && aggregationFilters.length > 0) {
    const transformedFilters = aggregationFiltersToParams(aggregationFilters);
    params.set('labels', JSON.stringify(transformedFilters.labels));
    params.set('values', JSON.stringify(transformedFilters.values));
  }
  if (token) params.set('token', token);
  return params;
};

/**
 * The URL Manager is responsible for synchronizing state between
 * CatalogDriver and the URL. There are 3 main cases we handle when
 * synchronizing:
 *
 * 1. When the app loads, CatalogDriver will need to
 * read the current state from the URL, in order to perform the search
 * expressed by the query string. `getStateFromURL` is used for this case.
 *
 * 2. When the URL changes as a result of `pushState` or `replaceState`,
 * CatalogDriver will need to be notified and given the updated state, so that
 * it can re-run the current search. `onURLStateChange` is used for this case.
 *
 * 3. When state changes internally in the CatalogDriver, as a result of an
 * Action, it will need to notify the URLManager of the change. `pushStateToURL`
 * is used for this case.
 *
 * Note that only CatalogDriver relevant query params are synchronized with URL.
 * For non-relevant query params, upon `pushStateToURL`, URL will be synchronized and
 * discard these params.
 */

export default class URLManager {
  private readonly _history;
  private _lastPushSearchString;
  private _unlisten!: VoidFunction;

  constructor({ history }: { history?: BrowserHistory } = {}) {
    this._history = history || createHistory();
    this._lastPushSearchString = '';
  }

  /**
   * Parse the current URL into application state
   *
   * @returns URL state
   */
  getStateFromURL(): Partial<CatalogDriverRequestURLState> {
    const params = new URLSearchParams(this._history.location.search);
    return paramsToState(params);
  }

  /**
   * Push the current state of the application to the URL
   *
   * @param state - the request state from the CatalogDriver
   * @param options - when pushing state to the URL, use history 'replace'
   * rather than 'push' to avoid adding a new history entry
   */
  pushStateToURL(
    state: CatalogDriverRequestURLState,
    { replaceUrl = false }: { replaceUrl?: boolean } = {}
  ) {
    const paramsToString = stateToParams(state).toString();
    this._lastPushSearchString = paramsToString;
    const navigationFunction = replaceUrl ? this._history.replace : this._history.push;
    navigationFunction({
      search: `?${paramsToString}`
    });
  }

  /**
   * Add an event handler to be executed whenever state is pushed to the URL
   *
   * @callback requestCallback
   * @param {Object} state - Updated application state parsed from the new URL
   *
   * @param {requestCallback} callback
   */
  onURLStateChange(callback: (state: Partial<CatalogDriverRequestURLState>) => void) {
    this._unlisten = this._history.listen(({ location }) => {
      // If this URL is updated as a result of a pushState request, we don't
      // want to notify that the URL changed.
      if (`?${this._lastPushSearchString}` === location.search) return;

      // Once we've decided to return based on lastPushSearchString, reset
      // it so that we don't break back / forward button.
      this._lastPushSearchString = '';

      callback(paramsToState(new URLSearchParams(location.search)));
    });
  }

  tearDown() {
    this._unlisten();
  }
}
