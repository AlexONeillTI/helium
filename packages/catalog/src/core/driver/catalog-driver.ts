import CatalogDriverBase from './catalog-driver-base';
import { CatalogDriverConfig } from './types';
import * as actions from './actions';

/**
 * The CatalogDriver is a framework agnostic catalog state manager
 */
export default class CatalogDriver extends CatalogDriverBase {
  private readonly _actions;
  constructor(config: CatalogDriverConfig) {
    super(config);

    this._actions = Object.entries(actions).reduce(
      (acc, [actionName, action]) => ({
        ...acc,
        [actionName]: action.bind(this)
      }),
      {}
    ) as actions.CatalogDriverActions;
  }

  getActions() {
    return this._actions;
  }
}
