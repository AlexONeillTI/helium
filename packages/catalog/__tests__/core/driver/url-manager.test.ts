import URLManager from '../../../src/core/driver/url-manager';
import { DEFAULT_STATE } from '../../../src/core/driver/catalog-driver-base';
import { createMemoryHistory } from 'history';

const setup = () => {
  const mockHistory = createMemoryHistory();
  const manager = new URLManager({ history: mockHistory });
  return {
    manager,
    mockHistory
  };
};

const baseUrlState = {
  page: 2,
  aggregationFilters: [{ label: 'label', value: 'value' }],
  token: 'abc',
  searchTerm: 'test'
};

const baseState = {
  ...DEFAULT_STATE,
  ...baseUrlState
};

const baseStateAsUrl =
  '?page=2&query=test&labels=%5B%22label%22%5D&values=%5B%22value%22%5D&token=abc';

describe('@thoughtindustries/catalog/CatalogDriver#URLManager', () => {
  it('should create new instance', () => {
    const urlManager = new URLManager();
    expect(urlManager).toBeInstanceOf(URLManager);
  });

  describe('getStateFromURL', () => {
    it('should parse the current state from the URL', () => {
      const { manager, mockHistory } = setup();
      mockHistory.push(
        '/?labels=%5B%22label1%22%2C%22label2%22%5D&values=%5B%22value1%22%2C%22value2%22%5D&page=3&query=test&token=abc'
      );

      const state = manager.getStateFromURL();
      expect(state).toMatchSnapshot();
    });

    it('should ignore unrecognized parameters', () => {
      const { manager, mockHistory } = setup();
      mockHistory.push('/?query=test&foo=bar&anything');

      const state = manager.getStateFromURL();
      expect(state).toMatchSnapshot();
    });

    it('should correctly handle multiple instances of the same parameter (take first instance)', () => {
      const { manager, mockHistory } = setup();
      mockHistory.push(
        '/?query=test&query=bad&token=abc&token=badtoken&page=3&page=1&labels=%5B%22label%22%5D&labels=%5B%22badlabel%22%5D&values=%5B%22value%22%5D&values=%5B%22badvalue%22%5D'
      );

      const state = manager.getStateFromURL();
      expect(state).toMatchSnapshot();
    });
  });

  describe('pushStateToURL', () => {
    it('should push the url from state', () => {
      const { manager, mockHistory } = setup();
      const spyPush = jest.spyOn(mockHistory, 'push');
      const spyReplace = jest.spyOn(mockHistory, 'replace');
      manager.pushStateToURL(baseState);
      expect(spyPush).toHaveBeenCalledWith({
        search: baseStateAsUrl
      });
      expect(spyReplace).toHaveBeenCalledTimes(0);
    });

    it('should replace the url from state', () => {
      const { manager, mockHistory } = setup();
      const spyPush = jest.spyOn(mockHistory, 'push');
      const spyReplace = jest.spyOn(mockHistory, 'replace');
      manager.pushStateToURL(baseState, { replaceUrl: true });
      expect(spyPush).toHaveBeenCalledTimes(0);
      expect(spyReplace).toHaveBeenCalledWith({
        search: baseStateAsUrl
      });
    });
  });

  describe('onURLStateChange', () => {
    it('should call provided callback with updated state when url changes', () => {
      const { manager, mockHistory } = setup();
      let newState;

      // Provide url state change handler
      manager.onURLStateChange(state => {
        newState = state;
      });

      // change url location
      mockHistory.push(baseStateAsUrl);

      // Verify it is called when url state changes
      expect(newState).toEqual(expect.objectContaining(baseUrlState));
    });

    /*
      If we triggered the callback every time we pushed state to the url,
      we would have an infinite loop.

      pushStateToURL -> UpdateUrl -> HistoryEvent -> Handler -> pushstateToURL

      Instead, we just want:
      pushStateToURL -> UpdateUrl -> HistoryEvent -> Handler

    */
    it('should not trigger callback as a result of pushStateToURL', () => {
      const { manager } = setup();
      let newState;

      // Provide url state change handler
      manager.onURLStateChange(state => {
        newState = state;
      });

      // Push state to the url
      manager.pushStateToURL(baseState);

      // Verify handler is not called
      expect(newState).toBe(undefined);
    });

    it('should trigger callback if navigating back to a url with browser buttons', () => {
      const { manager, mockHistory } = setup();
      let newState;

      // Provide url state change handler
      manager.onURLStateChange(state => {
        newState = state;
      });

      // Push state to the url
      manager.pushStateToURL(baseState);

      // Back button to default state
      mockHistory.back();
      expect(newState).toEqual({});

      // Forward button to double back to the original "newUrl"
      mockHistory.forward();
      expect(newState).toEqual(expect.objectContaining(baseUrlState));
    });
  });
});
