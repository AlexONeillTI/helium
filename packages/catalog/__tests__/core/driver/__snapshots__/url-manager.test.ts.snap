// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@thoughtindustries/catalog/CatalogDriver#URLManager getStateFromURL should correctly handle multiple instances of the same parameter (take first instance) 1`] = `
Object {
  "aggregationFilters": Array [
    Object {
      "label": "label",
      "value": "value",
    },
  ],
  "page": 3,
  "searchTerm": "test",
  "token": "abc",
}
`;

exports[`@thoughtindustries/catalog/CatalogDriver#URLManager getStateFromURL should ignore unrecognized parameters 1`] = `
Object {
  "searchTerm": "test",
}
`;

exports[`@thoughtindustries/catalog/CatalogDriver#URLManager getStateFromURL should parse the current state from the URL 1`] = `
Object {
  "aggregationFilters": Array [
    Object {
      "label": "label1",
      "value": "value1",
    },
    Object {
      "label": "label2",
      "value": "value2",
    },
  ],
  "page": 3,
  "searchTerm": "test",
  "token": "abc",
}
`;
